lib/Sema/CSSimplify.cpp

```diff
@@ -1908,6 +1908,12 @@ ConstraintSystem::matchFunctionTypes(FunctionType *func1, FunctionType *func2,
     return getTypeMatchFailure(locator);
   }
 
+  if (Context.LangOpts.UseClangFunctionTypes) {
```

Comment:
It would be great if this was a constraint system option instead which is set in `typeCheckExpression`.

Reply:
Is this something which has been added recently or is it a source breaking change?

Reply:
> It would be great if this was a constraint system option instead which is set in typeCheckExpression.

I could add an extra flag to `ConstraintSystemFlags`, and use that, but I can't remove the flag off `LangOptions` because it's being used in a bunch of other places which don't have access to a `ConstraintSystem`.

Should I still add an extra flag to `ConstraintSystemFlags`?

---

> Is this something which has been added recently or is it a source breaking change?

The change has been in the works for a while. This is probably the oldest PR: https://github.com/apple/swift/pull/27479. 

`-use-clang-function-types` is a frontend flag. While having the flag turned on is technically a source breaking change, in most cases it should be non-breaking, since people don't use `@convention(c)` functions directly (and when they do, the types generally line up). That said, the [GRDB library in the source compat suite does hit some breakage](https://github.com/apple/swift/pull/34914#issuecomment-737384068), which is why the feature flag is not turned on yet.

I'm going to investigate adding an implicit conversion for a subset of cases (where the Clang type mismatch is innocuous) after this PR, to avoid that breakage.

Reply:
bq. I could add an extra flag to ConstraintSystemFlags, and use that, but I can't remove the flag off LangOptions.

It's okay to keep LangOpts one, just pass it to the ConstraintSystem and it would be fine.

Reply:
Sounds good regarding the flag - if it's turned off by default I'd consider it to be a non-breaking change although I'd be great if we did have an `experimental` section to make it clear that source compatibility doesn't matter at this point.

Reply:
Fixed by adding a new case to `ConstraintSystemFlags` and setting it based on the language option.